proposals/defer-hydration.md

```diff
@@ -0,0 +1,90 @@
+
+# Overview
+
+The Defer Hydration Protocol specifies an attribute named `defer-hydration` that is placed on elements, usually during server rendering, to tell them not to hydrate when they are upgraded. Removing this attribute is a signal that the element should hydrate. Elements can observe the attribute being removed via `observedAttribute` and `attributeChangedCallback`.
+
+When an element hydrates it can remove the `defer-hydration` attribute from its shadow children to hydrate them, or keep the attribute if itself can determine a more optimal time to hydrate all or certain children. By making the parent responsible for removing the `defer-hydration` attribute from it's children, we ensure top-down ordering.
+
+## Use case 1: Auto-hydration with top-down odering
+
+In this use case we want to page to hydrate as soon as elements are defined, but we want to force top-down ordering to avoid invalid child states. Here we configure the server-rendering step to add the `defer-hydration` attribute to all elements _except_ the top-most defer-hydration-aware elements in the document.
+
+When the top-most elements are defined, they will run their hydrations steps since they don't have a `defer-hydration` attribute, and will trigger their subtrees to hydrate by removing `defer-hydration` from children.
+
+Example HTML:
+
+```html
+<!doctype html>
+<html>
+  <head>
+    <script type="module" src="./app.js"></script>
+  </head>
+  <body>
+    <x-nav>
+      <template shadowroot="open">
+        <x-header defer-hydration>
+          <template shadowroot="open">
+            <h1>Example</h1>
+          </template>
+        </header>
+      </template>
+    <x-article>
+      <template shadowroot="open">
+        <x-figure defer-hydration>...</x-figure>
+      </template>
+    </x-article>
+  </body>
+<html>
+```
+
+## Use case 2: On-demand hydration
+
+Hydration can be deferred until some data or user-driven signal, such as interacting with an element. In this case server-rendering is configured to add `defer-hydration` to all elements so that nothing will automatically hydrate.
+
+An app-level coordinator may implement an event delegation/buffering/replay system to detect user-events within an element and remove `defer-hydration` on demand before replaying events.
+
+*TODO: do we need to have an event that signals that hydration is complete before replaying events?*
+
+## Hydrating children
+
+```ts
+class MyElement extends HTMLElement {
+  static observedAttributes = ['defer-hydration'];
+
+  attributeChangedCallback(name, oldValue, newValue) {
+    if (name === 'defer-hydration' && newValue === null) {
+      this._hydrate();
+    }
+  }
+
+  _hydrate() {
+    // do template hydrate work
+    // ...
```

Comment:
One thing that's not clear to me from this proposal: what does "hydration" mean? Just `connectedCallback`? Or `constructor` plus `connectedCallback`?

I'm assuming that it only means `connectedCallback`, but you could imagine that a web component could have some logic in its `constructor` that assumes client-side rendering, which assumes top-down rendering, and thus breaks when `customElements.define` is called out-of-order.

Is this situation considered out-of-scope for this proposal?

Reply:
Thanks for the clarification. It might be good to nail down some SHOULDs in this case (not necessarily MUSTs). E.g., thoughts that pop into my head are:

- Should a component-author-author'ed (as opposed to framework-author-authored) `connectedCallback` fire in a deferred way?
- What about framework-specific callbacks related to reactivity/rendering? E.g. in LWC we have `renderedCallback` for when the DOM is rendered, Lit has various callbacks like `updated`/`firstUpdated`/etc.
- What if somehow a component is disconnected before the deferred `connectedCallback` code fires? Should `disconnectedCallback` fire before `connectedCallback`, or not at all?